### PR TITLE
Pytest + Speed improvements in unit tests.

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Attempting to speed up the unit tests to improve developer feedback.

Changes:
- Switch to using pytests and having the output collected by CircleCI. We get lots of convenient meta data.
- Fix a bug in the TGA tutorial that accidentally ignored parameters; reduced some others.
- Delete a teacher that was 100% redundant with another. Deleted the one that was less used. Update the corresponding tests.
- A few things can be switched to evaluating pretrained models instead of training new ones and be safe imho.
- Catch a bunch of tests that were calling parse_args without explicit parameters so that they can be safe.
- Moved a few things around to be better compatible with pytests
- Added explicit empty command line args to some scripts, so the test versions do not actually use sys.argv.
- Update relevant READMEs.

**Testing steps**
CI timing.

### CODE REVIEW

**Comment by stephenroller** [comment]
With parallelism:

image

Okay that wasn't the easy win I was hoping. Still 20 minutes (bottlenecked by at least one test). Compared to the parent without parallelism:

image

So we don't even get any real benefits. We'll need to profile the tests and find the offenders.

**Comment by stephenroller** [comment]
Following some code I found online on how to benchmark tests, here's all tests that take longer than 10s. A few are real offenders.

```
(1.55e+02s) test_generation (test_examples.TestExampleSeq2Seq)
(19.9s) test_pyt_batchsort_field (test_pytorch_data_teacher.TestPytorchDataTeacher)
(20.7s) test_pyt_batchsort_train (test_pytorch_data_teacher.TestPytorchDataTeacher)
(20.3s) test_pyt_batchsort_train_stream (test_pytorch_data_teacher.TestPytorchDataTeacher)
(10.7s) test_pyt_preprocess_train (test_pytorch_data_teacher.TestPytorchDataTeacher)
(21.2s) test_pyt_train (test_pytorch_data_teacher.TestPytorchDataTeacher)
(22.3s) test_pyt_train_stream (test_pytorch_data_teacher.TestPytorchDataTeacher)
(26.0s) test_valid_pyt_batchsort (test_pytorch_data_teacher.TestPytorchDataTeacher)
(56.0s) test_legacy_version (test_torch_agent.TestLegacyVersioning)
(17.6s) test_eval_fixed (test_tra.TestPolyRanker)
(14.7s) test_eval_fixed_label_not_in_cands (test_tra.TestPolyRanker)
(15.7s) test_eval_fixed (test_tra.TestTransformerRanker)
(34.3s) test_invsqrt_learning_rate (test_transformers.TestLearningRateScheduler)
(13.5s) test_beamsearch_blocking (test_transformers.TestTransformerGenerator)
(36.7s) test_beamsearch_contextblocking (test_transformers.TestTransformerGenerator)
(14.6s) test_generator_backcomp (test_transformers.TestTransformerGenerator)
```

**Comment by stephenroller** [comment]
Those add up to 499s on my devfair, with the full suite taking 11:15, so cleaning up these would be a big improvement.

**Comment by stephenroller** [comment]
The invsqrt learning rate one was taking forever because of --skip-generation and evaluating the entire valid set.

**Comment by stephenroller** [comment]
Since i got test timing running on circleci:
```
(11.7s) test_pyt_batchsort_train_stream (test_pytorch_data_teacher.TestPytorchDataTeacher)
(62.8s) test_pyt_preprocess_train (test_pytorch_data_teacher.TestPytorchDataTeacher)
(11.2s) test_pyt_train (test_pytorch_data_teacher.TestPytorchDataTeacher)
(49.8s) test_pyt_train_stream (test_pytorch_data_teacher.TestPytorchDataTeacher)
(20.1s) test_pyt_train_stream_ordered (test_pytorch_data_teacher.TestPytorchDataTeacher)
(30.2s) test_valid_pyt_batchsort (test_pytorch_data_teacher.TestPytorchDataTeacher)
(11.0s) test_beamsearch (test_seq2seq.TestSeq2Seq)
(45.3s) test_generation (test_seq2seq.TestSeq2Seq)
(12.4s) test_ranking (test_seq2seq.TestSeq2Seq)
(95.7s) test_legacy_version (test_torch_agent.TestLegacyVersioning)
(37.8s) test_eval_fixed (test_tra.TestMemNN)
(36.3s) test_eval_fixed (test_tra.TestPolyRanker)
(28.6s) test_eval_fixed_label_not_in_cands (test_tra.TestPolyRanker)
(42.2s) test_eval_fixed (test_tra.TestTransformerRanker)
(10.0s) test_badinput (test_transformers.TestTransformerGenerator)
(21.0s) test_beamsearch (test_transformers.TestTransformerGenerator)
(58.6s) test_beamsearch_blocking (test_transformers.TestTransformerGenerator)
(34.8s) test_beamsearch_contextblocking (test_transformers.TestTransformerGenerator)
(19.1s) test_generator_backcomp (test_transformers.TestTransformerGenerator)
(22.5s) test_greedysearch (test_transformers.TestTransformerGenerator)
(18.4s) test_nucleus (test_transformers.TestTransformerGenerator)
(22.5s) test_topk (test_transformers.TestTransformerGenerator)
(22.6s) test_xlm (test_transformers.TestTransformerGenerator)
(12.9s) test_alt_reduction (test_transformers.TestTransformerRanker)
(30.0s) test_repeater (test_transformers.TestTransformerRanker)
```

File: tests/test_transformers.py [hunk 1]
```diff
@@ -851,4 +856,4 @@ def test_invsqrt_learning_rate(self):
 
 
 if __name__ == '__main__':
-    unittest.main()
+    unittest.main(testRunner=testing_utils.TimeLoggingTestRunner)
```
Comment (stephenroller) [inline tests/test_transformers.py]: drop this

**Comment by stephenroller** [comment]
```
Slow Tests (>10.0s):
(26.6s) test_generation (test_examples.TestExampleSeq2Seq)
(21.7s) test_pyt_batchsort_field (test_pytorch_data_teacher.TestPytorchDataTeacher)
(55.9s) test_pyt_batchsort_train (test_pytorch_data_teacher.TestPytorchDataTeacher)
(51.3s) test_pyt_batchsort_train_preprocess (test_pytorch_data_teacher.TestPytorchDataTeacher)
(13.5s) test_pyt_batchsort_train_stream (test_pytorch_data_teacher.TestPytorchDataTeacher)
(59.0s) test_pyt_preprocess_train (test_pytorch_data_teacher.TestPytorchDataTeacher)
(42.2s) test_pyt_train (test_pytorch_data_teacher.TestPytorchDataTeacher)
(56.5s) test_pyt_train_stream (test_pytorch_data_teacher.TestPytorchDataTeacher)
(68.6s) test_pyt_train_stream_ordered (test_pytorch_data_teacher.TestPytorchDataTeacher)
(48.7s) test_valid_pyt_batchsort (test_pytorch_data_teacher.TestPytorchDataTeacher)
(13.7s) test_ranking (test_seq2seq.TestSeq2Seq)
(29.1s) test_eval_fixed (test_tra.TestMemNN)
(64.6s) test_eval_fixed (test_tra.TestPolyRanker)
(32.9s) test_eval_fixed_label_not_in_cands (test_tra.TestPolyRanker)
(33.0s) test_eval_fixed (test_tra.TestTransformerRanker)
(11.5s) test_beamsearch (test_transformers.TestTransformerGenerator)
(26.7s) test_beamsearch_blocking (test_transformers.TestTransformerGenerator)
(24.2s) test_beamsearch_contextblocking (test_transformers.TestTransformerGenerator)
(17.1s) test_generator_backcomp (test_transformers.TestTransformerGenerator)
(12.6s) test_greedysearch (test_transformers.TestTransformerGenerator)
(13.5s) test_xlm (test_transformers.TestTransformerGenerator)
(12.6s) test_alt_reduction (test_transformers.TestTransformerRanker)
(15.0s) test_repeater (test_transformers.TestTransformerRanker)
```

**Comment by stephenroller** [comment]
image

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -12,9 +12,9 @@ We actively welcome your pull requests.
 2. If you've added code that should be tested, add tests.
 3. If you've changed APIs, update the documentation.
 4. Autoformat and lint your code (`bash autoformat.sh`)
-5. Ensure the test suite passes. Run `python setup.py test`.
+5. Ensure the test suite passes. Run `python -m pytest -m unit`.
 6. If you've added a new dataset, you should also run
-   `python setup.py test -s tests.suites.datatests`. Copy-paste the output into a
+   `python -m pytest -m datatests`. Copy-paste the output into a
```
Comment (stephenroller) [inline CONTRIBUTING.md]: should be `-m data`

File: docs/source/tutorial_torch_generator_agent.md [hunk 3]
```diff
@@ -1,16 +1,16 @@
 # Using Torch Generator Agent
 
-**Authors**: Eric Smith
+Authors: Eric Smith
```
Comment (stephenroller) [inline docs/source/tutorial_torch_generator_agent.md]: This was to fight my editor, can be undone.

File: parlai/utils/testing.py [hunk 4]
```diff
@@ -16,7 +16,9 @@
 import shutil
 import io
 import signal
+import time
```
Comment (stephenroller) [inline parlai/utils/testing.py]: delete

File: parlai/utils/testing.py [hunk 5]
```diff
@@ -16,7 +16,9 @@
...
 from typing import Tuple
+from unittest.runner import TextTestResult
```
Comment (stephenroller) [inline parlai/utils/testing.py]: delete

File: tests/test_interative.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env python
```
Comment (stephenroller) [inline tests/test_interative.py]: drop from PR.

Comment (stephenroller) [reply tests/test_interative.py]: eh, it's fine.

File: tests/nightly/cpu/test_convai2_cpu.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import unittest
+import parlai.utils.testing as testing_utils
+
+
+"""
+This module ensures that the baseline models for convai2 produce the
+correct ~~validation~~ results.
```
Comment (emilydinan) [inline tests/nightly/cpu/test_convai2_cpu.py]: what is with the ~~

File: .circleci/config.yml [hunk 8]
```diff
@@ -320,8 +339,10 @@ jobs:
             - "~/venv/lib"
       - run:
           name: All mturk tests
-          command: coverage run setup.py test -s tests.suites.mturk -v
+          command: coverage run -m pytest --junitxml=test-results/junit.xml -m mturk parlai/mturk/core/test/*.py
```
[the remark this answers is not in the document]
Comment (stephenroller) [reply .circleci/config.yml]: Jack didn't like it.

File: parlai/agents/examples/seq2seq.py [hunk 9]
```diff
@@ -48,10 +48,12 @@ def __init__(self, embeddings, hidden_size):
         # must call super on all nn.Modules.
         super().__init__()
 
-        _vocab_size, esz = embeddings.weight.shape
         self.embeddings = embeddings
         self.lstm = nn.LSTM(
-            input_size=esz, hidden_size=hidden_size, num_layers=1, batch_first=True
+            input_size=hidden_size,
```
[the remark this answers is not in the document]
Comment (stephenroller) [reply parlai/agents/examples/seq2seq.py]: Yes, otherwise we need an extra layer to project from hidden-size to embedding-size for the final softmax. As this is just a tutorial, this sort of simplification is fine imho.

File: parlai/core/build_data.py [hunk 10]
```diff
@@ -416,7 +416,15 @@ def modelzoo_path(datapath, path):
...
         except (ImportError, AttributeError):
-            pass
+            try:
+                # maybe we didn't find a specific model, let's try generic .build
+                animal_ = '.'.join(animal.split(".")[:-1]) + '.build'
+                module_name = 'parlai.zoo.{}'.format(animal_)
+                my_module = importlib.import_module(module_name)
+                my_module.download(datapath)
+            except (ImportError, AttributeError):
+                # truly give up
```
[the remark this answers is not in the document]
Comment (stephenroller) [reply parlai/core/build_data.py]: We could. Current code doesn't and I'm open to making it a thing.

File: tests/nightly/cpu/test_convai2_cpu.py (hunk 7 above)
Comment (stephenroller) [reply tests/nightly/cpu/test_convai2_cpu.py]: idk, this was just there. All i did was rename this file.

File: parlai/tasks/integration_tests/agents.py [hunk 11]
```diff
@@ -284,27 +284,6 @@ class DefaultDataset(CandidateTeacherDataset):
     pass
 
 
-class MultipassTeacher(CandidateTeacher):
```
[the remark this answers is not in the document]
Comment (stephenroller) [reply parlai/tasks/integration_tests/agents.py]: It was redundant with the multiturn teacher, see PR description.

File: parlai/agents/examples/seq2seq.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (emilydinan) [reply parlai/agents/examples/seq2seq.py]: sounds good

File: parlai/core/build_data.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (emilydinan) [reply parlai/core/build_data.py]: i think we should because i have entered the model file path wrong before and then the error is obscured
